store/src/java/com/zimbra/cs/account/auth/AuthMechanism.java

```diff
@@ -249,13 +249,21 @@ public void doAuth(LdapProv prov, Domain domain, Account acct, String password,
                 if (PasswordUtil.SSHA512.verifySSHA512(encodedPassword, password)) {
                     return; // good password, RETURN
                 }  else {
+                    acct.refreshUserCredentials();
+                    if (PasswordUtil.SSHA512.verifySSHA512(acct.getAttr(Provisioning.A_userPassword), password)) {
```

Comment:
Can the password change whether is is SSHA or SSHA512?  If so, we should be able to cope with that.

Reply:
I don't think it can in between deciding it's of type SSHA512 and the user updating it in the background.

Reply:
Perhaps someone who knows more can say definitively.  My suspicion is that passwords generated by old versions of Zimbra might theoretically still be SSHA but new passwords would always be SSHA512.

Reply:
Ok.  Wouldn't the second case be more likely to run into this situation?
i.e. the old password was of type SSHA and the new updated password is of type SSHA512?

Reply:
Yeah, I guess so.

Reply:
As Gren noted, this is off (and apparently has been for a while).  There's no no guarantee the new password shares the same encoding with the old one.  This applies for the SSHA512 and SSHA cases.  Heck, the new password could be PLAIN... sigh.  So, it feels like we've over optimized, and can fail in a less than ideal way.

For each of the else cases, we should probably just do ```acct.refreshUserCredentials();``` if we're optimistic that our cache is stale.  But, even then, perhaps we should fall through to authenticating to LDAP as per the non-optimized case and the next time a request comes in we'll have the updated cache.

If we're not optimistic, perhaps we don't bother doing a manual refresh of the cache, but just let it age out normally ~15m and just always fall through regardless on the case where we fail to positively match the hashes.